[Navigation] Abort ongoing navigation on window.stop()

rwlbuis opened the pull request:
#### 92b5ddfd057326297847ab105fe65e50437ad65b
<pre>
[Navigation] Abort ongoing navigation on window.stop()
<a href="https://bugs.webkit.org/show_bug.cgi?id=275659">https://bugs.webkit.org/show_bug.cgi?id=275659</a>

Reviewed by Darin Adler and Alex Christensen.

Implement the &quot;inform the navigation API about aborting navigation&quot; alogithm [1] and
call it as part of the &quot;stop loading&quot; algorithm [2].

[1] <a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#inform-the-navigation-api-about-aborting-navigation">https://html.spec.whatwg.org/multipage/nav-history-apis.html#inform-the-navigation-api-about-aborting-navigation</a>
[2] <a href="https://html.spec.whatwg.org/multipage/document-lifecycle.html#nav-stop">https://html.spec.whatwg.org/multipage/document-lifecycle.html#nav-stop</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/signal-abort-detach-in-onnavigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/signal-abort-window-stop-in-onnavigate-expected.txt:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::stopAllLoaders):
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::innerDispatchNavigateEvent):
(WebCore::Navigation::abortOngoingNavigationIfNeeded):
* Source/WebCore/page/Navigation.h:

Canonical link: <a href="https://commits.webkit.org/280526@main">https://commits.webkit.org/280526@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3b3e2d94c11aa0c9a4a1bae285680c8d0cf10ed7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56887 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36215 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9361 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60506 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7330 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59015 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43838 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7520 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46076 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5144 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58917 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34016 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49097 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26934 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30796 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6428 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6334 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52756 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6699 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62188 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/800 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/6802 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53332 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/56579 "Build is in progress. Recent messages:") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/803 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49152 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53368 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12573 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/678 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32044 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33129 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34214 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32875 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2020,6 +2020,10 @@ void FrameLoader::stopAllLoaders(ClearProvisionalItem clearProvisionalItem, Stop
             localChild->checkedLoader()->stopAllLoaders(clearProvisionalItem);
     }
 
+    RefPtr window = m_frame->document()->protectedWindow();
```
Seems like this should be inside the body of the `if` statement below.

rwlbuis replied:
Good point, done.

darinadler commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -795,4 +794,12 @@ bool Navigation::dispatchDownloadNavigateEvent(const URL& url, const String& dow
 }
 
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#inform-the-navigation-api-about-aborting-navigation
+void Navigation::informAboutCanceledNavigation()
```
This is sort of in the “uncanny valley” of function names, similar to, but not the same as, the name in the specification, and without an object, the phrase “inform about” isn’t so great. I think we can do a little better with the name. One way is to just write it all out, `informNavigationAPIAboutAbortingNavigation`. Or we could just leave it named `abortOngoingNavigation` and remove the argument from that function? Or overload `abortOngoingNavigation` to have this be a version without an argument? Or name it `abortOngoingNavigateEvent`? Maybe there are some even better names.

rwlbuis replied:
Note that in the spec the algorithm is a bit more complicated than in this implementation, since it has some logic for  the window event loop, which does not seem to be needed so far. Having said that, indeed the method name is quite generic and maybe a bit verbose. I'd rather leave the abortOngoingNavigation as-is, since it is not public. In the end I came up with abortOngoingNavigationIfNeeded, which I think is pretty clear, hope you like it.

rwlbuis committed:
[Navigation] Abort ongoing navigation on window.stop()
https://bugs.webkit.org/show_bug.cgi?id=275659

Reviewed by Darin Adler and Alex Christensen.

Implement the "inform the navigation API about aborting navigation" alogithm [1] and
call it as part of the "stop loading" algorithm [2].

[1] https://html.spec.whatwg.org/multipage/nav-history-apis.html#inform-the-navigation-api-about-aborting-navigation
[2] https://html.spec.whatwg.org/multipage/document-lifecycle.html#nav-stop

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/signal-abort-detach-in-onnavigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/signal-abort-window-stop-in-onnavigate-expected.txt:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::stopAllLoaders):
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::innerDispatchNavigateEvent):
(WebCore::Navigation::abortOngoingNavigationIfNeeded):
* Source/WebCore/page/Navigation.h:

Canonical link: https://commits.webkit.org/280526@main
